GH-103929: handle long input lines with PyOS_InputHook

tacaswell opened the pull request:
If:
 - stdin is in line buffer mode
 - the readline module is not loaded
 - not on windows
 - a GUI toolkit has installd PyOS_InputHook

Then, if the user enters lines longer than 98 characters into `input`:

 - user calls `input(...)`
 - user types/pastes a line longer than 98 characters + 1 newline
 - PyOS_InputHook returns
 - the first 99 characters are returned
 - the last character is not a new line so we go back to my_fgets
 - the PyOS_InputHook blocks because despite there being value in the buffer, stdin does not flag itself as ready to be read again
 - user hits enter again and to finish reading the input
 - the extra new line comes out the next time the user calls input

This fixes this bug by passing the currently read number of characters to `my_fgets` to skip the input hook when reading out a long buffer.

closes #103929


------

Qustions:

 - I do not know how to test this other than manually
 - It seems to work when stdin is unbuffered, but I do not fully understand why
 - Is the edit to the docs of PyOS_InputHook correct?


<!-- gh-issue-number: gh-103929 -->
* Issue: gh-103929
<!-- /gh-issue-number -->

tacaswell committed:
GH-103929: handle long input lines with PyOS_InputHook

If:
 - stdin is in line buffer mode
 - the readline module is not loaded
 - not on windows
 - a GUI toolkit has installd PyOS_InputHook

Then, if the user enters lines longer than 98 characters into `input`:

 - user calls `input(...)`
 - user types/pastes a line longer than 98 characters + 1 newline
 - PyOS_InputHook returns
 - the first 99 characters are returned
 - the last character is not a new line so we go back to my_fgets
 - the PyOS_InputHook blocks because despite there being value in the buffer,
   stdin does not flag itself as ready to be read again
 - user hits enter again and to finish reading the input
 - the extra new line comes out the next time the user calls input

This fixes this bug by passing the currently read number of characters to
`my_fgets` to skip the input hook when reading out a long buffer.

closes #103929

tacaswell committed:
GH-103929: handle long input lines with PyOS_InputHook

If:
 - stdin is in line buffer mode
 - the readline module is not loaded
 - not on windows
 - a GUI toolkit has installd PyOS_InputHook

Then, if the user enters lines longer than 98 characters into `input`:

 - user calls `input(...)`
 - user types/pastes a line longer than 98 characters + 1 newline
 - PyOS_InputHook returns
 - the first 99 characters are returned
 - the last character is not a new line so we go back to my_fgets
 - the PyOS_InputHook blocks because despite there being value in the buffer,
   stdin does not flag itself as ready to be read again
 - user hits enter again and to finish reading the input
 - the extra new line comes out the next time the user calls input

This fixes this bug by passing the currently read number of characters to
`my_fgets` to skip the input hook when reading out a long buffer.

closes #103929

mdboom commented:
@brandtbucher: Maybe you have enough context to review this since you've been in readline-adjacent things lately?

brandtbucher commented:
I'll take a look today. Thanks for the PR!

tacaswell commented:
rebased to handle conflicts with the sub-processor work.

tacaswell commented:
Rebased again.

jdtsmith commented:
@brandtbucher did you ever get a chance to review this one?

mdboom commented:
@pablogsal, @ambv: If I understand correctly, users are more likely to hit this bug in 3.14 since the new REPL doesn't use readline.  Should we try to get this into the next 3.14 beta?

pablogsal commented:
> @pablogsal, @ambv: If I understand correctly, users are more likely to hit this bug in 3.14 since the new REPL doesn't use readline. Should we try to get this into the next 3.14 beta?

This fix doesn't apply to the new REPL, so I don't think how this PR could help here

mdboom commented:
> > @pablogsal, @ambv: If I understand correctly, users are more likely to hit this bug in 3.14 since the new REPL doesn't use readline. Should we try to get this into the next 3.14 beta?
> 
> This fix doesn't apply to the new REPL, so I don't think how this PR could help here

Got it. My bad.

jdtsmith commented:
@pablogsal Does that mean the new REPL should/will not have this 100 char limit while `PyOS_InputHook` is installed?

tacaswell commented:
The bug that is fixed is in the readline-replacement implementation that Python carries, not in the input hooks, so if there are bugs in the new repl with long lines, they will be different bugs ;)

jdtsmith commented:
I'll take that as a solid maybe ;). Will report if I find similar long-line bugs with active input hooks.

mdehoon commented:
This bug shows up not only with `input()` but also with interactive Python when typing commands at the Python prompt.

For example, running the commands in this script one-by-one will freeze Python after `long_string` until the user hits enter one more time:
```python
>>> import tkinter
>>> t = tkinter.Tk()
>>> short_string = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
>>> long_string = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

>>> >>> ^D
```
(as @tacaswell  mentioned, the bug only appears on Python without readline).

jdtsmith commented:
It would be nice to get this reviewed and merged.  For now the only workaround is `import readline`, which may be undesirable.

tacaswell commented:
rebased to re-trigger CI.

tacaswell commented:
I think we need the inputhook inside of that `do...while` loop as if the input hook returns early (like some flavors on windows do).

Maybe it should be rephrased to:

> This function [PyOS_InputHook] should block until stdin is readable but may return early


This lets you run a loop like "run UI for 5ms, check stid, run UI for 5ms ...".  This bit of code is here due to https://github.com/python/cpython/pull/7978/files which I put in with guidance from @zooba .

tacaswell committed:
DOC: add additional qualifier to expected behavior of PyOS_InputHook

tacaswell committed:
DOC: add additional qualifier to expected behavior of PyOS_InputHook

tacaswell commented:
But does it happen inside the function for all function currently in the wild?  I have a recent-ish survey here: https://github.com/python/cpython/pull/119843#issuecomment-2143514240

mdehoon commented:
> But does it happen inside the function for all function currently in the wild? I have a recent-ish survey here: https://github.com/python/cpython/pull/119843#issuecomment-2143514240

In the main packages (matplotlib, PyQt5, PyQt6, PySide, and PyGTK), PyOS_InputHook points to a function that returns once data is available on stdin.

PyGObject (which superseded PyGTK) does not use PyOS_InputHook.

For wxPython: there was a discussion about it 16 years ago (see https://discuss.wxpython.org/t/interactive-usage-of-wxpython/32128/22, where I was making the same comments). But I did not find any (recent or old) version of wxPython that actually uses PyOS_InputHook, so I guess it was never implemented.

Now there may be some module in the wild that assumes that PyOS_InputHook is called repeatedly. I didn't find such a module, but if it exists, its event loop will freeze with Python without readline support.

tacaswell committed:
MNT: move the PyOS_InputHook call up a loop level

If we have returned from the input hook then there is something to read in the
stdin so read it out to the new line before calling the input hook again.

tacaswell committed:
MNT: move the PyOS_InputHook call up a loop level

If we have returned from the input hook then there is something to read in the
stdin so read it out to the new line before calling the input hook again.

tacaswell commented:
@mdehoon is correct and we can pull the PyOS_InputHook up a level rather than add code to avoid calling it in the inner loop.

tacaswell commented:
@pablogsal With the new pyrepl what is the future plan for this code?

mdehoon commented:
@tacaswell  Thanks, it looks good to me.

tacaswell commented:
This is still waiting on review from a core developer.  This is to my knowledge still both correct and needed.

pablogsal committed:
Fix review comments for GH-103931

pablogsal commented:
I pushed a small fix for this in `codex/pr-103931-fix-96fa029507a6`. Can you take a look?

tacaswell commented:
Merged that commit to this branch.  Sorry, thought this was below the threshold of needing a news.

tacaswell commented:
@pablogsal I think this is good to go.
